File: typescript/src/message.ts
/*
 * Copyright (C) 2024 The Noisy Sockets Authors.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

export const APIVersion: string = "nsh/v1alpha1";

interface Meta {
  // apiVersion is the version of the API.
  apiVersion: string;
  // kind is the kind of message.
  kind: string;
  // id is the globally unique message ID, or in the case of an Ack,
  // the ID of the message being acknowledged.
  id?: string;
}

// AckStatus is the status of an acknowledgment.
export const enum AckStatus {
  // OK indicates the message was processed successfully.
  OK = "OK",
  // ERROR indicates an error occurred while processing the message.
  ERROR = "ERROR",
  // UNIMPLEMENTED indicates the message was not processed because the
  // feature / message type is not implemented.
  UNIMPLEMENTED = "UNIMPLEMENTED",
}

// Ack acknowledges a message has been received and processed.
interface Ack extends Meta {
  // status is the status of the acknowledgment.
  status: AckStatus;
  // reason is the reason for the status, typically populated for errors.
  reason: string;
}

// Data is a message containing terminal data.
interface Data extends Meta {
  // data is the base64 encoded terminal data.
  data: string;
}

// TerminalOpenRequest requests a new terminal session.
interface TerminalOpenRequest extends Meta {
  // columns is the number of columns in the terminal.
  columns: number;
  // rows is the number of rows in the terminal.
  rows: number;
  // env is a list of environment variables to pass to the shell process.
  env: string[];
}

// TerminalExit indicates the terminal session / shell has exited.
interface TerminalExit extends Meta {
  // status is the exit status of the shell process.
  status: number;
}

// TerminalResize resizes the terminal window.
interface TerminalResize extends Meta {
  // columns is the number of columns in the terminal.
  columns: number;
  // rows is the number of rows in the terminal.
  rows: number;
}

// MaxDataMessageBytes is the maximum number of bytes that can be sent in a
// single data message. This has been cribbed from the Go implementation.
export const MaxDataMessageBytes: number = 49089;

export type Message =
  | Ack
  | Data
  | TerminalOpenRequest
  | TerminalExit
  | TerminalResize;

export type {
  Meta,
  Ack,
  Data,
  TerminalOpenRequest,
  TerminalExit,
  TerminalResize,
};

File: internal/message/message.go
// SPDX-License-Identifier: MPL-2.0
/*
 * Copyright (C) 2024 The Noisy Sockets Authors.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package message

import (
	"encoding/json"
	"fmt"

	"github.com/mitchellh/mapstructure"
)

const (
	// MaxSize is the maximum supported WebSocket message size.
	MaxSize = 65536
)

type Meta struct {
	// APIVersion is the version of the API.
	APIVersion string `json:"apiVersion" mapstructure:"apiVersion"`
	// Kind is the kind of the message.
	Kind string `json:"kind" mapstructure:"kind"`
	// ID is the globally unique message ID, or in the case of an Ack, the ID of
	// the message being acknowledged.
	ID string `json:"id" mapstructure:"id"`
}

func (m *Meta) GetID() string {
	return m.ID
}

func (m *Meta) SetID(id string) {
	m.ID = id
}

// Message is a message that can be sent/received.
type Message interface {
	GetAPIVersion() string
	GetKind() string
	GetID() string
	SetID(id string)
}

var knownTypes = make(map[string]func() Message)

// RegisterType registers a known message type.
func RegisterType(newMessageOfType func() Message) {
	msg := newMessageOfType()
	knownTypes[Type(msg)] = newMessageOfType
}

// Marshal marshals a message, returning the data.
func Marshal(msg Message) ([]byte, error) {
	// Convert the message to a map.
	m := make(map[string]interface{})
	if err := mapstructure.Decode(msg, &m); err != nil {
		return nil, err
	}

	// Add the message type information to the map.
	m["apiVersion"] = msg.GetAPIVersion()
	m["kind"] = msg.GetKind()

	// Marshal the map to JSON.
	data, err := json.Marshal(m)
	if err != nil {
		return nil, err
	}

	return data, nil
}

// Unmarshal unmarshals a message from the given data.
func Unmarshal(data []byte) (Message, error) {
	var meta Meta
	if err := json.Unmarshal(data, &meta); err != nil {
		return nil, err
	}

	typ := fmt.Sprintf("%s/%s", meta.APIVersion, meta.Kind)
	newMessageOfType, ok := knownTypes[typ]
	if !ok {
		return nil, fmt.Errorf("unknown message type: %s", typ)
	}

	msg := newMessageOfType()
	if err := json.Unmarshal(data, msg); err != nil {
		return nil, err
	}

	return msg, nil
}

// Type returns the string formatted type of the message.
func Type(msg Message) string {
	return fmt.Sprintf("%s/%s", msg.GetAPIVersion(), msg.GetKind())
}
